Support Globalnet EgressIPs at the cluster level

sridhargaddam opened the pull request:
This PR includes support for ClusterGlobalEgressIPs.
Currently, the code is validated manually and using
the unit tests but will be enabled in a future PR
once the other controllers are also implemented.

Fixes issue: https://github.com/submariner-io/submariner/issues/1163
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

sridhargaddam committed:
Support Globalnet EgressIPs at the cluster level

This PR includes support for ClusterGlobalEgressIPs.
Currently, the code is validated manually and using
the unit tests but will be enabled in a future PR
once the other controllers are also implemented.

Fixes issue: https://github.com/submariner-io/submariner/issues/1163
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -70,13 +74,6 @@ func NewClusterGlobalEgressIPController(config syncer.ResourceSyncerConfig, pool
-			return nil, err
-		}
-	}
-
```
This was done on startup so we can ensure each object will be able to reserve its previously allocated IPs (unless the global CIDR changed). Otherwise, in doing this in the callback, we may process a newly created object first which could allocate a previously used block, in which case we'd need to reallocate.

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+	// If numGlobalIPs is modified, delete the existing allocation.
+	if len(status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(status)
+		_ = c.pool.Release(status.AllocatedIPs...)
```
We should report the error.

sridhargaddam replied:
The only error from c.pool.Release() is when an invalid IP is passed as input which should not happen in an ideal case. So, I intentionally did not log it. 

But anyways, I can log the message.

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+func (c *clusterGlobalEgressIPController) onRemove(key string, egressIP *submarinerv1.ClusterGlobalEgressIP) bool { // nolint unparam
+	if len(egressIP.Status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(&egressIP.Status)
+		_ = c.pool.Release(egressIP.Status.AllocatedIPs...)
```
We should report the error.

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -93,32 +90,52 @@ func NewClusterGlobalEgressIPController(config syncer.ResourceSyncerConfig, pool
+	prevStatus := clusterGlobalEgressIP.Status
 
-	klog.Infof("Processing %sd %#v", op, clusterGlobalEgressIP)
+	if err := c.validate(numberOfIPs, clusterGlobalEgressIP); err != nil {
+		klog.Warningf("Error: %v", err)
+		return checkGlobalEgressIPStatusChanged(&prevStatus, &clusterGlobalEgressIP.Status, clusterGlobalEgressIP), false
+	}
```
I don't think we want to do this validation for delete - we can't update the status anyway. Perhaps we shouldn't do anything for delete?

Also `allocateIPs` already validates the numberOfIPs.

 I think we should handle create and update the same, ie `onCreateOrUpdate`.  On a restart, if the object was changed while we were down, then it's the same as an update.

sridhargaddam replied:
> I don't think we want to do this validation for delete - we can't update the status anyway. Perhaps we shouldn't do anything for delete?

Yes, we can remove validate for delete.

> I think we should handle create and update the same, ie `onCreateOrUpdate`. On a restart, if the object was changed while we were down, then it's the same as an update.

Done

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+
+func (c *clusterGlobalEgressIPController) flushClusterGlobalEgressRules(status *submarinerv1.GlobalEgressIPStatus) {
+	snatIP := c.getTargetSNATIPaddress(status.AllocatedIPs)
+	if snatIP == "" {
+		klog.Warning("flushClusterGlobalEgressRules called with 0 AllocatedIPs")
+		return
+	}
```
I think it's reasonable to assume this is only called with `AllocatedIPs` > 0.  This would be a code path that is never hit. Same on L252.

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+
+	if allocatedIPs == 1 {
+		snatIP = allocIPs[0]
+	} else if allocatedIPs > 1 {
+		snatIP = fmt.Sprintf("%s-%s", allocIPs[0], allocIPs[len(allocIPs)-1])
+	}
```
Again, let's assume > 0 to avoid an unused code path.

```suggestion
	} else {
		snatIP = fmt.Sprintf("%s-%s", allocIPs[0], allocIPs[len(allocIPs)-1])
	}
```

tpantelis commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+	return snatIP
+}
+
+func (c *clusterGlobalEgressIPController) allocateGlobalIPs(key string, numberOfIPs *int, status *submarinerv1.GlobalEgressIPStatus) bool {
```
I had a general `allocateIPs` function reused for cluster and global egress IP. Is that problematic with how the IP tables stuff is done?

sridhargaddam replied:
Yes Tom. The IPTable rules will be different for the two. IMO its better to have separate APIs. We can consider optimizing it later after we implement the support for egress IPs.

tpantelis commented on pkg/globalnet/controllers/iptablesdriver/iface.go:
```diff
@@ -0,0 +1,73 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package iptablesdriver
```
```suggestion
package iptables
```

tpantelis commented on pkg/globalnet/controllers/iptablesdriver/iface.go:
```diff
@@ -0,0 +1,73 @@
+	RemoveClusterEgressRules(sourceIP, snatIP, globalNetIPTableMark string) error
+}
+
+type IPTableIface struct {
```
```suggestion
type ipTables struct {
```

tpantelis commented on pkg/globalnet/controllers/iptablesdriver/iface.go:
```diff
@@ -0,0 +1,73 @@
+	ipt iptables.Interface
+}
+
+func NewIPTableInterface() (*IPTableIface, error) {
```
```suggestion
func New() (Interface, error) {
```

tpantelis commented:
@sridhargaddam  How about the following for abstracting the IP table rules for pod egress, cluster egress, and ingress:

```
type Interface interface {
	AddRules(sourceIP, snatIP, globalNetIPTableMark string) error
	RemoveRules(sourceIP, snatIP, globalNetIPTableMark string) error
}

func ForPodEgress() (Interface, error) ...
func ForClusterEgress() (Interface, error) ...
func ForIngress() (Interface, error) ...
```

I'm assuming common API methods can be defined. `globalNetIPTableMark` is a constant so I think we can remove it from the API.

Each controller would create its own instance and and it can be  stored in the `baseIPAllocationController`.  Then perhaps we can reuse IP allocation code between the controllers.

Does this look feasible?

tpantelis commented:
> @sridhargaddam How about the following for abstracting the IP table rules for pod egress, cluster egress, and ingress:
> ...
> Does this look feasible?

I pushed https://github.com/submariner-io/submariner/pull/1397. Hopefully this looks reasonable and can at least be a starting point.

mangelajo commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -93,32 +90,52 @@ func NewClusterGlobalEgressIPController(config syncer.ResourceSyncerConfig, pool
 
 func (c *clusterGlobalEgressIPController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
 	clusterGlobalEgressIP := from.(*submarinerv1.ClusterGlobalEgressIP)
+	numberOfIPs := clusterGlobalEgressIP.Spec.NumberOfIPs
+	if numberOfIPs == nil {
+		one := 1
+		numberOfIPs = &one
+	}
```
```suggestion
	numberOfIPs := 1
	if clusterGlobalEgressIP.Spec.NumberOfIPs != nil {
		numberOfIPs = *clusterGlobalEgressIP.Spec.NumberOfIPs
	}
```

And from this point on you can avoid passing down pointers, because I was reading the other functions and suffering because numberOfIps could be nil, although you set it to &one here ;)

mangelajo commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+		tryAppendStatusCondition(&egressIP.Status, &metav1.Condition{
+			Type:    string(submarinerv1.GlobalEgressIPAllocated),
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidInput",
```
We should make those Reason values part of the submarinerv1 module, as constants, so then the users of the API can compare to those constant values when consuming the API.

sridhargaddam replied:
Yeah, makes sense. WDYS @tpantelis

tpantelis replied:
yeah we can although not sure what other users of our Go API we would have.  The Reason is really a string for end-users.

mangelajo commented on pkg/globalnet/controllers/cluster_egressip_controller.go:
```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+			return false
+		}
+
+		if err := c.programClusterGlobalEgressRules(status); err != nil {
+			_ = c.pool.Release(status.AllocatedIPs...)
+			status.AllocatedIPs = []string{}
+
+			klog.Errorf("Error syncing the IPTable rules on the node for %q: %v", key, err)
+
+			return true
```
I wonder if at some point we should make those entitites separate:
1) The one handling allocations & reporting to the object
2) The one configuring egress rules, because it listenst to  GlobalEgressIPs

mangelajo commented on pkg/globalnet/controllers/cluster_egressip_controller_test.go:
```diff
@@ -35,7 +37,15 @@ var _ = Describe("ClusterGlobalEgressIP controller", func() {
 			t.awaitClusterGlobalEgressIPStatusAllocated(controllers.ClusterGlobalEgressIPName, 1)
-			// TODO - verify IP tables
+			// Verify that IPtable rule is programmed in SM-GN-EGRESS-CLUSTER chain with allocatedIPs
+			t.ipt.AwaitRule("nat", constants.SmGlobalnetEgressChainForCluster,
```
This is covering functional testing which we should probably keep on a different category, and even different execution to unit tests. For example, I'm not sure if we could execute those tests properly on an OSX environment.

Can we separate this type of system-interacting test to a separate file at least

cluster_egressip_controller_functional_test.go

And may be prepend this type of test with:

Describe("[functional] ClusterGlobalEgressIP controller", func()

So later we could... either move them to a separate $project/test/functional or filter them out in ginkgo?

sridhargaddam replied:
As you probably know, we use fake iptables (pkg/iptables/fake) during the unit testing. We were doing similar validation even with the existing globalnet. I hope we did not find any issue running on OSX (please correct me if this is not the case). Basically, if we did not have any issue so far on OSX, we should not have any issue even with this PR as well.

sridhargaddam committed:
Address possible race condition for existing objects

When the cluster already has a ClusterGlobalEgressIP object and
Globalnet pod gets restarted, we should reserve the allocatedIPs
before we start processing any new objects. This PR fixes it.

Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam reviewed (commented):
Addresed the review comments @tpantelis

sridhargaddam reviewed (commented):
Thanks for reviewing @mangelajo

sridhargaddam committed:
Modified numGlobalIPs field from pointer to int

Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam commented:
> @sridhargaddam How about the following for abstracting the IP table rules for pod egress, cluster egress, and ingress:
> 
> ```
> type Interface interface {
> 	AddRules(sourceIP, snatIP, globalNetIPTableMark string) error
> 	RemoveRules(sourceIP, snatIP, globalNetIPTableMark string) error
> }
> 
> func ForPodEgress() (Interface, error) ...
> func ForClusterEgress() (Interface, error) ...
> func ForIngress() (Interface, error) ...
> ```
> 
> I'm assuming common API methods can be defined. `globalNetIPTableMark` is a constant so I think we can remove it from the API.
> 
> Each controller would create its own instance and and it can be stored in the `baseIPAllocationController`. Then perhaps we can reuse IP allocation code between the controllers.
> 
> Does this look feasible?

Should be possible to do this. But IMHO lets first get the basic functionality for datapath working and we can look at this as an optimization once we enable the new controllers.
